### ISSUE TITLE
preped recipe requires 1000x more space when saved as a RDS from within a function than from "plain code"

### ISSUE THREAD

**Issue by SimonCoulombe** [tidymodels/recipes#533]
The code below creates 2 .RDS files.
"temp.rds"  is created from plain code and requires 5 KB.
"temp2.rds" is created from the same code wrapped inside a function and requires 5  284 KB

does anyone have any idea?  This code is inspired from this issue https://github.com/tidymodels/recipes/issues/301

```
library(rlang)
  library(recipes)
  library(dplyr)

big_iris <- bind_rows(rep_along(1:1000, list(as_tibble(iris))))

rec_small <- recipe(Sepal.Length ~ Species + Sepal.Width, big_iris[0,]) %>%
  step_center(Sepal.Width)

p_rec_small <- prep(rec_small, big_iris, retain = FALSE)
readr::write_rds(p_rec_small, "temp.rds") # 5kb

myfunction <- function(){
  big_iris <- bind_rows(rep_along(1:1000, list(as_tibble(iris))))

  rec_small <- recipe(Sepal.Length ~ Species + Sepal.Width, big_iris[0,]) %>%
    step_center(Sepal.Width)
  
  p_rec_small <- prep(rec_small, big_iris, retain = FALSE)

  readr::write_rds(p_rec_small, "temp2.rds") # 5284kb  BIG
  
}

myfunction()
```

**Comment by topepo** [comment]
**tl;dr**: R objects can contain embedded environments that blow up the saved object size. The [`butcher`](https://tidymodels.github.io/butcher/index.html) package fixes that.  

**longer answer** 

A few types of R objects save the entire environment that the were called in. Here's an example for a basic formula:

``` r
f <- y ~ x
str(f)
#> Class 'formula'  language y ~ x
#>   ..- attr(*, ".Environment")=<environment: R_GlobalEnv>
```

<sup>Created on 2020-06-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Depending on how you save it, the environment is brought along. If you have some huge object in that environment that size is larger than expected. `object.size()` does not account for this but `lobstr::obj_size()` does. 

Recipes use _quosures_ to capture the terms that you select in each step and this includes these environment. 

``` r
a <- 2
foo <- function() rlang::quo(a)
foo()
#> <quosure>
#> expr: ^a
#> env:  0x7fa26fa79a38
```

<sup>Created on 2020-06-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

The [`butcher`](https://tidymodels.github.io/butcher/index.html) package was created to eliminate components of objects that will minimally affect the objects usage. The `butcher()` function will get rid of the quosure environments and other stuff. You can still use `bake()`, `juice()`, and `tidy()` on the butchered objects but we can't guarantee that `prep()` will still work. That's why we don't do this automatically. 

For your example: 

``` r
library(rlang)
library(recipes)
#> Loading required package: dplyr
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
#> 
#> Attaching package: 'recipes'
#> The following object is masked from 'package:stats':
#> 
#>     step
library(dplyr)
library(butcher)
library(lobstr)
```

``` r

big_iris <- bind_rows(rep_along(1:1000, list(as_tibble(iris))))

rec_small <- recipe(Sepal.Length ~ Species + Sepal.Width, big_iris[0,]) %>%
  step_center(Sepal.Width)

p_rec_small <- prep(rec_small, big_iris, retain = FALSE)

lobstr::obj_size(p_rec_small)
#> 14,032 B
weigh(p_rec_small)
#> # A tibble: 41 x 2
#>    object                       size
#>    <chr>                       <dbl>
#>  1 steps.terms              0.00559 
#>  2 template.Species         0.000648
#>  3 steps.means              0.000288
#>  4 var_info.variable        0.000264
#>  5 term_info.variable       0.000264
#>  6 levels.Species.values    0.000264
#>  7 orig_lvls.Species.values 0.000264
#>  8 last_term_info.variable  0.000264
#>  9 var_info.role            0.0002  
#> 10 term_info.role           0.0002  
#> # … with 31 more rows


myfunction <- function(){
  big_iris <- bind_rows(rep_along(1:1000, list(as_tibble(iris))))
  
  rec_small <- recipe(Sepal.Length ~ Species + Sepal.Width, big_iris[0,]) %>%
    step_center(Sepal.Width)
  
  p_rec_small <- prep(rec_small, big_iris, retain = FALSE)

  cat("-- Before butchering ---\n")
  cat("size: ", lobstr::obj_size(p_rec_small), "\n")
  print(weigh(p_rec_small))

  cat("\n\n-- After butchering ---\n")
  
  butchered <- butcher(p_rec_small, verbose = TRUE)
  cat("size: ", lobstr::obj_size(butchered), "\n")
  print(weigh(butchered))
}

myfunction()
#> -- Before butchering ---
#> size:  14032 
#> # A tibble: 41 x 2
#>    object                       size
#>    <chr>                       <dbl>
#>  1 steps.terms              5.42    
#>  2 template.Species         0.000648
#>  3 steps.means              0.000288
#>  4 var_info.variable        0.000264
#>  5 term_info.variable       0.000264
#>  6 levels.Species.values    0.000264
#>  7 orig_lvls.Species.values 0.000264
#>  8 last_term_info.variable  0.000264
#>  9 var_info.role            0.0002  
#> 10 term_info.role           0.0002  
#> # … with 31 more rows
#> 
#> 
#> -- After butchering ---
#> ✓ Memory released: '5,403,664 B'
#> size:  12376 
#> # A tibble: 41 x 2
#>    object                       size
#>    <chr>                       <dbl>
#>  1 template.Species         0.000648
#>  2 steps.terms              0.000624
#>  3 steps.means              0.000288
#>  4 var_info.variable        0.000264
#>  5 term_info.variable       0.000264
#>  6 levels.Species.values    0.000264
#>  7 orig_lvls.Species.values 0.000264
#>  8 last_term_info.variable  0.000264
#>  9 var_info.role            0.0002  
#> 10 term_info.role           0.0002  
#> # … with 31 more rows
```

<sup>Created on 2020-06-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
